prow/plank/reconciler.go

```diff
@@ -0,0 +1,645 @@
+			pj.SetComplete()
+			pj.Status.State = prowv1.FailureState
+			pj.Status.Description = "Job failed."
+
+		case corev1.PodPending:
+			maxPodPending := r.config().Plank.PodPendingTimeout.Duration
+			maxPodUnscheduled := r.config().Plank.PodUnscheduledTimeout.Duration
+			if pod.Status.StartTime.IsZero() {
+				if time.Since(pod.CreationTimestamp.Time) >= maxPodUnscheduled {
+					// Pod is stuck in unscheduled state longer than maxPodUncheduled
+					// abort the job, and talk to GitHub
+					pj.SetComplete()
+					pj.Status.State = prowv1.ErrorState
+					pj.Status.Description = "Pod scheduling timeout."
+					r.log.WithFields(pjutil.ProwJobFields(pj)).Info("Marked job for stale unscheduled pod as errored.")
+					break
+				}
+			} else if time.Since(pod.Status.StartTime.Time) >= maxPodPending {
+				// Pod is stuck in pending state longer than maxPodPending
+				// abort the job, and talk to GitHub
+				pj.SetComplete()
+				pj.Status.State = prowv1.ErrorState
+				pj.Status.Description = "Pod pending timeout."
+				r.log.WithFields(pjutil.ProwJobFields(pj)).Info("Marked job for stale pending pod as errored.")
+				break
+			}
+			// Pod is running. Do nothing.
+			return nil
+		case corev1.PodRunning:
+			maxPodRunning := r.config().Plank.PodRunningTimeout.Duration
+			if pod.Status.StartTime.IsZero() || time.Since(pod.Status.StartTime.Time) < maxPodRunning {
+				// Pod is still running. Do nothing.
+				return nil
+			}
+
+			// Pod is stuck in running state longer than maxPodRunning
+			// abort the job, and talk to GitHub
+			pj.SetComplete()
+			pj.Status.State = prowv1.AbortedState
+			pj.Status.Description = "Pod running timeout."
+			client, ok := r.buildClients[pj.ClusterAlias()]
+			if !ok {
+				return fmt.Errorf("running pod %s: unknown cluster alias %q", pod.Name, pj.ClusterAlias())
+			}
+			if err := client.Delete(r.ctx, pod); err != nil {
+				return fmt.Errorf("failed to delete pod %s that was in running timeout: %v", pod.Name, err)
+			}
+			r.log.WithFields(pjutil.ProwJobFields(pj)).Info("Deleted stale running pod.")
+		default:
+			// other states, ignore
+			return nil
+		}
+	}
+
+	pj.Status.URL, err = pjutil.JobURL(r.config().Plank, *pj, r.log)
+	if err != nil {
+		r.log.WithFields(pjutil.ProwJobFields(pj)).WithError(err).Warn("failed to get jobURL")
+	}
+
+	if prevPJ.Status.State != pj.Status.State {
+		r.log.WithFields(pjutil.ProwJobFields(pj)).
+			WithField("from", prevPJ.Status.State).
+			WithField("to", pj.Status.State).Info("Transitioning states.")
+	}
+
+	if err := r.pjClient.Patch(r.ctx, pj.DeepCopy(), ctrlruntimeclient.MergeFrom(prevPJ)); err != nil {
+		return fmt.Errorf("patching prowjob: %v", err)
+	}
+
+	return nil
+}
+
+func (r *reconciler) syncTriggeredJob(pj *prowv1.ProwJob) (*reconcile.Result, error) {
+	prevPJ := pj.DeepCopy()
+
+	var id, pn string
+
+	pod, podExists, err := r.pod(pj)
+	if err != nil {
+		return nil, err
+	}
+	// We may end up in a state where the pod exists but the prowjob is not
```

Comment:
nit: this massive comment block is for the `else` ... consider flipping the logic

Reply:
Done